src/network/networkd-json.c

```diff
@@ -509,6 +509,106 @@ static int dns_append_json(Link *link, JsonVariant **v) {
         return json_variant_set_field_non_null(v, "DNS", array);
 }
 
+static int dnr_append_json_one(Link *link, const struct sd_dns_resolver *res, NetworkConfigSource s, const union in_addr_union *p, JsonVariant **array) {
+        _cleanup_(json_variant_unrefp) JsonVariant *addrs_array = NULL;
+        _cleanup_strv_free_ char **transports = NULL;
+        assert(link);
+        assert(res);
+        assert(array);
+
+        FOREACH_ARRAY(addr, res->addrs, res->n_addrs) {
+                json_variant_append_arrayb(
+                                &addrs_array,
+                                JSON_BUILD_IN_ADDR(addr, res->family));
+        }
+
+        sd_dns_resolver_transports_to_strv(res->transports, &transports);
+
+        //FIXME ifindex?
+        return json_variant_append_arrayb(
+                        array,
+                        JSON_BUILD_OBJECT(
+                                        JSON_BUILD_PAIR_INTEGER("Family", res->family),
+                                        JSON_BUILD_PAIR_INTEGER("Priority", res->priority),
+                                        JSON_BUILD_PAIR_VARIANT_NON_NULL("Addresses", addrs_array),
+                                        JSON_BUILD_PAIR_UNSIGNED_NON_ZERO("Port", res->port),
+                                        JSON_BUILD_PAIR_STRING_NON_EMPTY("ServerName", res->auth_name),
+                                        JSON_BUILD_PAIR_STRING_NON_EMPTY("DoHPath", res->dohpath),
+                                        JSON_BUILD_PAIR_STRV_NON_EMPTY("Transports", transports),
+                                        JSON_BUILD_PAIR_STRING("ConfigSource", network_config_source_to_string(s)),
+                                        JSON_BUILD_PAIR_IN_ADDR_NON_NULL("ConfigProvider", p, res->family)));
+}
+
+static int dnr_append_json(Link *link, JsonVariant **v) {
+        _cleanup_(json_variant_unrefp) JsonVariant *array = NULL;
+        int r;
+
+        assert(link);
+        assert(v);
+
+        if (!link->network)
+                return 0;
+
+        if (link->dhcp_lease && link_get_use_dnr(link, NETWORK_CONFIG_SOURCE_DHCP4)) {
+                struct sd_dns_resolver *dnr;
+                union in_addr_union s;
+                int n_dnr;
+
+                r = sd_dhcp_lease_get_server_identifier(link->dhcp_lease, &s.in);
+                if (r < 0)
+                        return r;
+
+                n_dnr = sd_dhcp_lease_get_dnr(link->dhcp_lease, &dnr);
```

Comment:
this can fal, no?

Reply:
Looks like it's kind of broken in the sense that I have quite a few DNRs received via router advertisements but `networktcl status --json` doesn't show them because it fails here:
```sh
561                     n_dnr = sd_dhcp_lease_get_dnr(link->dhcp_lease, &dnr);
(gdb) n
562                     if (n_dnr < 0)
(gdb) p n_dnr
$1 = -61
```
```sh
# cat /run/systemd/netif/state 
# This is private data. Do not parse.
OPER_STATE=routable
CARRIER_STATE=carrier
ADDRESS_STATE=routable
IPV4_ADDRESS_STATE=routable
IPV6_ADDRESS_STATE=degraded
ONLINE_STATE=online
DNS=1.1.1.1 2001:db8::1234#xbc.\124a\125 2001:db8::1234#xyz.\001\000\140 2001:db8:80::5678#abc.xy\250 2001:db8::9abc#abc.xy\250 2001:db8::5678#\255\255\255.abc\003xyz 2001:db8::9abe#\255\255\255.abc\003xyz 2001:db8::5678#abc.xyz 2001:db8::9abc#abc.xyz 2001:db8::1234#d\145\139.\001\000\. 2001:db8::1232#xyz.abc 2001:db8::1234#x\249z.abc 2001:db8::1234#x\.z.\001\000\. 2001:db8::1234#xyz.\.b\039 2001:db8::5678#a\033c.\136\134\250 2001:db8::9abc#a\033c.\136\134\250 2001:db8::1234#\.\.z.\001\000\. 2001:db8::1234#xyz.ab\039 2001:db8::5678#a\033k.\136\134\250 2001:db8::9abc#a\033k.\136\134\250 2001:db8:ff:fd00::5678#abc.\198\134\250 2001:db8::9abc#abc.\198\134\250 2001:db8::1234#xyz.abc 2001:db8::1230#xyz.abc 2001:db8::1234#d\145\139.\001\000\038 2001:db8::1234#x\125\124.abc 2001:db8::1234#\001\000\003.\.b\039 2001:db8::2c00:0:0:5678#a\033c.\136\134\250 2001:db8:ceff:66cc::9abc#a\033c.\136\134\250 2001:db8::5678#abc.xy\250 2001:db8::1234#xyz.\001\000c 2001:db8:0:5542:1c00:f050:0:1233#xyz.\001\000c dafe:f247::1233#xqz.ab\039 2001:db8::5678#abc.\136\134\250 2001:db8::9abc#abc.\136\134\250 2001:db8::1238#xyz.\001\000c 2001:db8::1734#xyz.\001\000c 2001:db8::1234#xyz.\001.c 20f7:cb8::1234#xyz.\001\000c 2001:db8::56:7820:10d:b800#abc.xyz ::9a:bc00:0:9abc#abc.xyz 2001:db8::1234#xyz-abc 2001:db8::1234#xyz\003abc 2001:db8::1234#xyz.\001\000\. dafe:f247::1232#xqz.ab\039 dafe:f247::1238#xqz.ab\039 2001:db8:0:94::1233#xyz.\001\000c 1ffe:db8::1234#xyz.\001\000c 2001:db8::1233#xyz.abc 2001:db8::1234#\001\000z.\.b\039 2001:db8::b:d496:650:5100#xyz.\001\000c 2001:db8::1235#xyz.\001\000c 2001:db8:ff:fd00::5678#\.bc.\198\134\250 2001:db8::9abc#\.bc.\198\134\250
```
```sh
# resolvectl 
Global
         Protocols: LLMNR=resolve -mDNS +DNSOverTLS DNSSEC=no/unsupported
  resolv.conf mode: stub

Link 2 (host0)
    Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
         Protocols: +DefaultRoute LLMNR=resolve -mDNS +DNSOverTLS DNSSEC=no/unsupported
       DNS Servers: 2001:db8::1234#xbc.124a125 2001:db8::1234#xyz.001000140 2001:db8:80::5678#abc.xy250 2001:db8::9abc#abc.xy250
                    2001:db8::5678#255255255.abc003xyz 2001:db8::9abe#255255255.abc003xyz 2001:db8::5678#abc.xyz 2001:db8::9abc#abc.xyz
                    2001:db8::1234#d145139.001000. 2001:db8::1232#xyz.abc 2001:db8::1234#x249z.abc 2001:db8::1234#x.z.001000. 2001:db8::1234#xyz..b039
                    2001:db8::5678#a033c.136134250 2001:db8::9abc#a033c.136134250 2001:db8::1234#..z.001000. 2001:db8::1234#xyz.ab039
                    2001:db8::5678#a033k.136134250 2001:db8::9abc#a033k.136134250 2001:db8:ff:fd00::5678#abc.198134250 2001:db8::9abc#abc.198134250
                    2001:db8::1234#xyz.abc 2001:db8::1230#xyz.abc 2001:db8::1234#d145139.001000038 2001:db8::1234#x125124.abc 2001:db8::1234#001000003..b039
                    2001:db8::2c00:0:0:5678#a033c.136134250 2001:db8:ceff:66cc::9abc#a033c.136134250 2001:db8::5678#abc.xy250 2001:db8::1234#xyz.001000c
                    2001:db8:0:5542:1c00:f050:0:1233#xyz.001000c dafe:f247::1233#xqz.ab039 2001:db8::5678#abc.136134250 2001:db8::9abc#abc.136134250
                    2001:db8::1238#xyz.001000c 2001:db8::1734#xyz.001000c 2001:db8::1234#xyz.001.c 20f7:cb8::1234#xyz.001000c
                    2001:db8::56:7820:10d:b800#abc.xyz ::9a:bc00:0:9abc#abc.xyz 2001:db8::1234#xyz-abc 2001:db8::1234#xyz003abc 2001:db8::1234#xyz.001000.
                    dafe:f247::1232#xqz.ab039 dafe:f247::1238#xqz.ab039 2001:db8:0:94::1233#xyz.001000c 1ffe:db8::1234#xyz.001000c 2001:db8::1233#xyz.abc
                    2001:db8::1234#001000z..b039 2001:db8::b:d496:650:5100#xyz.001000c 2001:db8::1235#xyz.001000c 2001:db8:ff:fd00::5678#.bc.198134250
                    2001:db8::9abc#.bc.198134250 1.1.1.1
     Default Route: yes
```